TailLight/device.cpp

```diff
@@ -1,50 +1,186 @@
+NTSTATUS PnpNotifyDeviceInterfaceChange(
+    _In_ PVOID pvNotificationStructure,
+    _Inout_opt_ PVOID pvContext) {
 
-    WDFDEVICE device = (WDFDEVICE)WdfTimerGetParentObject(Timer);
-    NT_ASSERTMSG("EvtSetBlackTimer device NULL\n", device);
+    //KdPrint(("TailLight: PnpNotifyDeviceInterfaceChange enter\n"));
+    NT_ASSERTMSG("WDFDEVICE not passed in!", pvContext);
 
-    NTSTATUS status = SetFeatureColor(device, 0);
-    if (!NT_SUCCESS(status)) {
-        KdPrint(("TailLight: EvtSetBlackTimer failure NTSTATUS=0x%x\n", status));
-        return;
+    if (pvNotificationStructure == NULL) {
+        return STATUS_SUCCESS;
     }
 
-    KdPrint(("TailLight: EvtSetBlackTimer end\n"));
+    PDEVICE_INTERFACE_CHANGE_NOTIFICATION pDevInterface =
+        (PDEVICE_INTERFACE_CHANGE_NOTIFICATION)pvNotificationStructure;
+
+    ASSERT(IsEqualGUID(*(_GUID*)&(pDevInterface->InterfaceClassGuid),
+        *(_GUID*)&GUID_DEVINTERFACE_HID));
+
+    {     
+        // Assumption: Device will arrive before removal.
+        if (IsEqualGUID(*(LPGUID)&(pDevInterface->Event),
+            *(LPGUID)&GUID_DEVICE_INTERFACE_ARRIVAL)) {
+
+            // Opening a device may trigger PnP operations. Ensure that either a
+            // timer or a work item is used when opening up a device.
+            // Refer to p356 of Oney and IoGetDeviceObjectPointer.
+            //
+            // NOTE: It is possible for us to get blocked waiting for a system
+            // thread. One solution would be to use a timer that spawns a
+            // system thread on timeout vs. a wait loop.
+
+
+            // Always must return success, so ignore result.
+            CreateWorkItemForIoTargetOpenDevice((WDFDEVICE)pvContext,
+                *pDevInterface->SymbolicLinkName);
+        }
+    }
+
+    return STATUS_SUCCESS;
 }
 
+
 NTSTATUS EvtSelfManagedIoInit(WDFDEVICE device) {
-    // Initialize tail-light to black to have control over HW state
-    WDF_TIMER_CONFIG timerCfg = {};
-    WDF_TIMER_CONFIG_INIT(&timerCfg, EvtSetBlackTimer);
 
-    WDF_OBJECT_ATTRIBUTES attribs = {};
-    WDF_OBJECT_ATTRIBUTES_INIT(&attribs);
-    attribs.ParentObject = device;
-    attribs.ExecutionLevel = WdfExecutionLevelPassive; // required to access HID functions
+    WDFDRIVER driver = WdfDeviceGetDriver(device);
+    PDRIVER_CONTEXT driverContext = WdfObjectGet_DRIVER_CONTEXT(driver);
+
+    //TRACE_FN_ENTRY;
+
+    NTSTATUS status = STATUS_SUCCESS;
+
+    status = IoRegisterPlugPlayNotification(
+        EventCategoryDeviceInterfaceChange,
+        PNPNOTIFY_DEVICE_INTERFACE_INCLUDE_EXISTING_INTERFACES,
+        (PVOID)&GUID_DEVINTERFACE_HID,
+        WdfDriverWdmGetDriverObject(driver),
+        PnpNotifyDeviceInterfaceChange,
+        (PVOID)device,
+        &driverContext->pnpDevInterfaceChangedHandle
+    );
+
+    //TRACE_FN_EXIT;
+
+    return status;
+}
+
+NTSTATUS AllocatePDOName(WDFDEVICE device, UNICODE_STRING& pdoName) 
```

Comment:
Why are you extracting `DevicePropertyPhysicalDeviceObjectName` retrieval into a separate function here? This doesn't seem to be required for this PR.

Reply:
Previous work environments had different code quality stages. The idea is that you should do the minimum amount of work to validate a design, then clean up the design afterward it is validated or throw away the code that fails. There has not been broad testing of this code so I submitted the minimum. Copy/paste has low risk and minimum effort.